musig/tests/musig_tests.rs

```diff
@@ -0,0 +1,184 @@
+mod musig_tests {
+    use bellman::pairing::bn256::Bn256;
+    use franklin_crypto::alt_babyjubjub::{AltJubjubBn256, FixedGenerators};
+    use franklin_crypto::eddsa::{PrivateKey, PublicKey, Seed};
+    use musig::musig::{MusigSession, MusigVerifier};
+    use musig::hash::Sha256HStar;
+    use rand::{thread_rng, Rng};
+
+    struct SplitIterator<'a, T> {
+        index: usize,
+        left: &'a mut [T],
+        right: &'a mut [T],
+    }
+
+    impl<'a, T> SplitIterator<'a, T> {
+        fn new(left: &'a mut [T], right: &'a mut [T]) -> SplitIterator<'a, T> {
+            SplitIterator {
+                index: 0,
+                left,
+                right,
+            }
+        }
+
+        fn next(&mut self) -> Option<&mut T> {
+            let res = if self.index < self.left.len() {
+                Some(&mut self.left[self.index])
+            } else if self.index < self.left.len() + self.right.len() {
+                Some(&mut self.right[self.index - self.left.len()])
+            } else {
+                None
+            };
+
+            self.index += 1;
+
+            res
+        }
+    }
+
+    fn split_slice_at_inclusive<T>(slice: &mut [T], mid: usize) -> (SplitIterator<T>, &T) {
+        let len = slice.len();
+        let ptr = slice.as_mut_ptr();
+
+        unsafe {
```

Comment:
>This code is used to borrow one element of a slice as mutable, and others as immutable. Do you know how to do this better?

I think it's an [XY problem](https://en.wikipedia.org/wiki/XY_problem). You don't really **have** to borrow mutable/immutable at the same time.

You can simply clone the data from the immutable element, and then iterate through other elements as mutable, skipping one of them, like:

```rust
let data = elemets[i].data().clone();

for (_, element) in elements.iter_mut().enumerate().filter(|(id, _el) id != i) {
    element.do_something(&data);
}
```

Reply:
Yes, cloning elements was an obvious solution for the problem, but I wanted to avoid that. Borrowing some elements of a slice as mutable and others as immutable is totally legit and doesn't break any rust ideology. It's just the compiler that cannot figure out at compilation time that those two sets of elements doesn't intersect and we don't produce more than 1 reference for any of the slice elements. Solving this case was fun, but ok, I can just use clone.

Reply:
Using `unsafe` in Rust is a sensitive topic. In general, there are two main reason (unrelated to environment, e.g. embedded) to use `unsafe`:

1. FFI.
2. Huge performance boost unachievable with safe code only.

Otherwise, it is highly recommended to avoid `unsafe`. Yeah, `unsafe` allows us to write code based on the data unavailable to compiler, but at the same time the **greatest** thing in Rust **is** the compiler, which makes you (not personally you, of course) worry about invariants of your code less. Each `unsafe` block in your code is a burden, which you have to carry. You'll have to maintain the invariants by your own, since you've told the compiler that you got it.

Thus, a small optimization here in my opinion does not overweight the introduced inconsistency.

If you are interested in topic, you can read about the usage of `unsafe` in `actix-web`: it was a big discussion (even close to holy war / flame) about `actix-web` library using `unsafe` without a strong motivation. While in most cases the crate maintainer definitely knew what he's doing, some of the code turned out to be unsound. At the same time, it has lead to a reputational damage, just because it's not conventional in Rust community to discard the guarantees (and indeed, without these guarantees, Rust is just *yet another language*).

Reply:
Everything related to iterations in std::slice is implemented using unsafe, so I don't see any problem in writing one more split function. What are other options? Copy data whenever you don't find corresponding standard function? No matter how big data is? Or create pull request to std lib and wait for approval?

Reply:
Code in `std` is allowed to be `unsafe` simply because standard library is under the constant supervision of the many qualified programmers, which work hard to ensure that every use of `unsafe` is correct and unavoidable.

I'm not saying that `unsafe` is an evil creature which is forbidden to be used. It **has** its usages and the performance is one of them (and this is stated in my comment).

However, performance of the test code is not the key priority, and if it's about choosing either of the unsafe code or data copying, I'd choose the latter.

P.S. Regarding

>  Copy data whenever you don't find corresponding standard function

I'm by no means an expert Rust programmer which has seen every possible case, but I've spent the last few years writing 95% of my code in Rust, both for the production and personal projects. However, cases where I clearly had to use `unsafe` are not numerous, and the majority of them were caused by FFI. So I don't really agree that it's always about either of `unsafe` or data copying when you can't find corresponding standard function.

Reply:
Fixed